File: apps/server/src/services/MettaParser.ts
import fs from 'fs';
import { MorkService } from './MorkService.js';



const morkUrl = process.env.MORK_URL || 'http://localhost:8001';
const morkService = new MorkService(morkUrl);

export interface Post {
  id: string;
  properties: Record<string, string>;
}

export class MettaParser {
  // Parses a metta file (needs pipeline to feed metta data) and returns structured posts
  static async parse(filePath: string, load = false): Promise<Post[]> {
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n').filter(line => line.trim() !== '');

    const posts: Record<string, Post> = {};
    const regex = /\(([\w-]+)\s+(A_\w+)\s+"([^"]+)"\)/;

    for (const line of lines) {
      if (load) { await morkService.uploadMetta(line); }
      const match = line.match(regex);
      if (match) {
        const [, property, id, value] = match;
        if (!posts[id]) {
          posts[id] = { id, properties: {} };
        }
        posts[id].properties[property] = value;
      }
    }
    return Object.values(posts);
  }

  // format the data to properly feed into embedding model, so that the embedding capture relevant data
  static toEmbeddingString(post: Post): string {
    return Object.entries(post.properties)
      .map(([key, value]) => `${key} ${post.id} "${value}"`)
      .join(' ');
  }
}

File: apps/server-py/services/metta_parser.py
import re
from dataclasses import dataclass, field

_LINE_REGEX = re.compile(r'\(([\w-]+)\s+(A_\w+)\s+"([^"]+)"\)')


@dataclass
class Post:
    id: str
    properties: dict = field(default_factory=dict)


def parse(file_path: str) -> list[Post]:
    with open(file_path, "r", encoding="utf-8") as f:
        lines = [line for line in f.read().split("\n") if line.strip() != ""]

    posts: dict[str, Post] = {}
    for line in lines:
        match = _LINE_REGEX.search(line)
        if match:
            prop, post_id, value = match.groups()
            if post_id not in posts:
                posts[post_id] = Post(id=post_id)
            posts[post_id].properties[prop] = value

    return list(posts.values())


def to_embedding_string(post: Post) -> str:
    return " ".join(
        f'{key} {post.id} "{value}"' for key, value in post.properties.items()
    )
